authz/providers/azure/rbac/rbac.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	u.clusterType = getClusterType(opts.AuthzMode)
+	u.lock = &sync.Mutex{}
+
+	return u, nil
+}
+
+func New(opts authzOpts.Options, authopts auth.Options, authzInfo *AuthzInfo) (*AccessInfo, error) {
+	rbacURL, err := url.Parse(authzInfo.ARMEndPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	var tokenProvider graph.TokenProvider
+	switch opts.AuthzMode {
+	case authzOpts.ARCAuthzMode:
+		tokenProvider = graph.NewClientCredentialTokenProvider(authopts.ClientID, authopts.ClientSecret,
+			fmt.Sprintf("%s%s/oauth2/v2.0/token", authzInfo.AADEndpoint, authopts.TenantID),
+			fmt.Sprintf("%s.default", authzInfo.ARMEndPoint))
+	case authzOpts.AKSAuthzMode:
+		tokenProvider = graph.NewAKSTokenProvider(opts.AKSAuthzURL, authopts.TenantID)
+	}
+
+	return newAccessInfo(tokenProvider, rbacURL, opts)
+}
+
+func (a *AccessInfo) RefreshToken() error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	if a.IsTokenExpired() {
+		resp, err := a.tokenProvider.Acquire("")
+		if err != nil {
+			glog.Errorf("%s failed to refresh token : %s", a.tokenProvider.Name(), err.Error())
+			return errors.Wrap(err, "failed to refresh rbac token")
+		}
+
+		// Set the authorization headers for future requests
+		a.headers.Set("Authorization", fmt.Sprintf("Bearer %s", resp.Token))
+		expIn := time.Duration(resp.Expires) * time.Second
+		a.expiresAt = time.Now().Add(expIn - expiryDelta)
+		glog.Infof("Token refreshed successfully on %s. Expire at:%s", time.Now(), a.expiresAt)
+	}
+
+	return nil
+}
+
+func (a *AccessInfo) IsTokenExpired() bool {
+	if a.expiresAt.Before(time.Now()) {
+		return true
+	} else {
+		return false
+	}
+}
+
+func (a *AccessInfo) ShouldSkipAuthzCheckForNonAADUsers() bool {
+	return a.skipAuthzForNonAADUsers
+}
+
+func (a *AccessInfo) GetResultFromCache(request *authzv1.SubjectAccessReviewSpec, store authz.Store) (bool, bool) {
+	var result bool
+	key := getResultCacheKey(request)
+	glog.V(10).Infof("Cache search for key: %s", key)
+	found, _ := store.Get(key, &result)
+	return found, result
+}
+
+func (a *AccessInfo) SkipAuthzCheck(request *authzv1.SubjectAccessReviewSpec) bool {
+	if a.clusterType == connectedClusters {
+		_, ok := a.skipCheck[strings.ToLower(request.User)]
+		return ok
+	}
+	return false
+}
+
+func (a *AccessInfo) SetResultInCache(request *authzv1.SubjectAccessReviewSpec, result bool, store authz.Store) error {
+	key := getResultCacheKey(request)
+	glog.V(10).Infof("Cache set for key: %s, value: %t", key, result)
+	return store.Set(key, result)
+}
+
+func (a *AccessInfo) AllowNonResPathDiscoveryAccess(request *authzv1.SubjectAccessReviewSpec) bool {
+	if request.NonResourceAttributes != nil && a.allowNonResDiscoveryPathAccess && strings.EqualFold(request.NonResourceAttributes.Verb, "get") {
+		path := strings.ToLower(request.NonResourceAttributes.Path)
+		if strings.HasPrefix(path, "/api") || strings.HasPrefix(path, "/openapi") || strings.HasPrefix(path, "/version") || strings.HasPrefix(path, "/healthz") {
```

Comment:
- What about `/apis` ?

```
 kubectl get clusterrole system:discovery -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  annotations:
    rbac.authorization.kubernetes.io/autoupdate: "true"
  creationTimestamp: "2020-05-31T20:46:20Z"
  labels:
    kubernetes.io/bootstrapping: rbac-defaults
  name: system:discovery
  resourceVersion: "43"
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterroles/system%3Adiscovery
  uid: c8118371-a37f-11ea-85c5-0242ac120002
rules:
- nonResourceURLs:
  - /api
  - /api/*
  - /apis
  - /apis/*
  - /healthz
  - /openapi
  - /openapi/*
  - /version
  - /version/
  verbs:
  - get
```

- Also, prefix checks should use a delimiter to make sure we are actually matching against a path component. So, these checks should match prefix against `/api/`, `/apis/` to accidentally matching against `/api123`, etc.

Reply:
Prefix /api covers all 4 cases. What I have observed is if user creates request on nonexistent non-res path /api123, request returns the right set of non-res paths as below:

http://127.0.0.1:8003/api123
{
  "paths": [
    "/apis",
    "/apis/",
    "/apis/apiextensions.k8s.io",
    "/apis/apiextensions.k8s.io/v1beta1",
    "/healthz",
    "/healthz/etcd",
    "/healthz/log",
    "/healthz/ping",
    "/healthz/poststarthook/crd-informer-synced",
    "/healthz/poststarthook/generic-apiserver-start-informers",
    "/healthz/poststarthook/start-apiextensions-controllers",
    "/healthz/poststarthook/start-apiextensions-informers",
    "/metrics",
    "/openapi/v2",
    "/version"
  ]

This behavior is consistent with K8s RBAC. Hence IMHO there should not be any issue in accidently matching against nonexistent non-res path. Do you see any escalation of privilege here?

Reply:
Using a delimiter for prefix checks just safer. But I think it is ok for current paths.